src/test/ui/unsafe/rfc-2585-unsafe_op_in_unsafe_fn.mir.stderr

```diff
@@ -76,12 +76,10 @@ LL |     unsafe {}
    |     ^^^^^^ unnecessary `unsafe` block
 
 error: unnecessary `unsafe` block
-  --> $DIR/rfc-2585-unsafe_op_in_unsafe_fn.rs:47:14
+  --> $DIR/rfc-2585-unsafe_op_in_unsafe_fn.rs:47:5
    |
 LL |     unsafe { unsafe { unsf() } }
-   |     ------   ^^^^^^ unnecessary `unsafe` block
-   |     |
-   |     because it's nested under this `unsafe` block
+   |     ^^^^^^ unnecessary `unsafe` block
```

Comment:
We might want to be nudging people towards larger unsafe blocks rather than to maximally small ones.

In unsafe code maintaining safety invariants will typically involve multiple independent statements, and at least some style guides suggest including both the unsafe operation itself as well any preceding statements that contribute to it being safe (e.g. a null check) into the same unsafe block.

Reply:
Well, I personally much prefer more fine-grained unsafe blocks. Otherwise it’s often quite hard to identify where _exactly_ the unsafe operations actually are and whether or not one might have overlooked one of them on review. AFAIK, there’s often no clear boundary as to which operations are and which aren’t relevant to maintaining safety invariants. And often even the behavior of “unrelated” safe operations (e.g. ones that can access private fields of a data structure you’re implementing) can be highly relevant for overall soundness.

Also the behavior as implemented makes for a straightforward solution of the problem that cases such as the one fixed in f3a599222b3b6b60b1f293a17100bb946839988d weren’t reported previously, or similarly cases such as the one I described [above](https://github.com/rust-lang/rust/pull/93678#issuecomment-1030685757):


> On current nightly,
> 
> ```rust
> unsafe fn unsf() {}
> 
> fn inner_ignored() {
>     unsafe {
>         #[allow(unused_unsafe)]
>         unsafe {
>             unsf()
>         }
>     }
> }
> ```
> 
> doesn’t create any warnings.

While the case with an inner `#[allow(unused_unsafe)]` is detectable, and actually that’s what I tried to do initially when working on this PR, for the external-macro case I’d need to find out how to detect those… maybe even non-external macros, too, because it always makes little sense to get a warning that suggests removing an `unsafe` block that’s coming from a macro. OTOH, special-casing warning-behavior on macros seems weird, I don’t know if there’s precedent (for non-external macros).

<hr>

Also, considering that while `unsf()` is unsafe Rust code, wrapping it in an `unsafe` block masks the unsafety, making `unsafe { unsf() }` a safe expression, so then it would make the most sense to me that in the expression `unsafe { unsafe { unsf() } }`, the outer `unsafe` is the unused one.

Really, this `unused_unsafe` lint combines two different kinds of lints, one for an `unsafe` that’s truly unused (as in, it only contains safe code, and `unsafe { unsf() }` is safe code), and another one for detecting `unsafe` blocks within other `unsafe` blocks, where the inner `unsafe` block is ***not*** unused but still *redundant*, because it’s appearing in a context where unsafe code would have been allowed. Of course, these two lints together can – if one is not careful – create bad overall suggestions (suggesting to remove too many `unsafe` blocks). The way to resolve this taken after this PR is to consider the warning about *actually* “unused” `unsafe` blocks to take precedent, and then the warning for redundant `unsafe` blocks within other `unsafe` blocks only triggers if the containing `unsafe` block turns out not to be unused. In my opinion it makes sense if a lint called `unused_unsafe` is mainly concerned with actually unused `unsafe` blocks.

Suggesting removal of the inner `unsafe` block in the expression `unsafe { unsafe { unsf() } }` instead would give precedent to the warning about the inner `unsafe` block being redundant and then, to avoid inconsistent suggestions/warnings, act as if the `unsf()` contained within that block actually used the outer `unsafe` block.

***So… well… of course, ultimately it stays question of mainly personal opinion on which kind of warning is preferred.***

Note that the case that one unused `unsafe` block contains *multiple* other/inner `unsafe` blocks is in an edit-distance kind of way closer to code with lots of small `unsafe` blocks only, compared to code with a single big `unsafe` block only. Warning only on the outer `unsafe` block creates fewer warnings and (implicitly) suggests fewer changes to the code. A piece of code that already perfectly contains an exhaustive set of smaller `unsafe` blocks (inside of one unused bigger one) seems to me as if the author intended to write lots of small `unsafe` blocks and accidentally added a large one around them, and it does not seems all that likely that the author intended to write one large `unsafe` blocks and accidentally (perfectly) managed to also enclose every single (smaller set of) unsafe operation(s) in that large `unsafe` block with its own dedicated `unsafe` block. Note that in any case where _not_ all unsafe operations have their own smaller `unsafe` block, the lint will still effectively nudge the user to using a single large `unsafe` block anyways (because there’s no good way to suggest adding extra `unsafe` blocks and also the lint (as implemented in this PR) only goes as far as to point to a containing `unsafe` block, and doesn’t also note where else that containing `unsafe` block is actually being used (although that might actually be a reasonable future addition)). The user would need to remove the outer `unsafe` block themself to find the places where it used to be used (directly).

Certainly, on that last point, the case where only a single inner `unsafe` block appears inside a single outer `unsafe` block is a bit of a special case, because then the edit-distance for removing either the outer or the inner one becomes the same; but special-casing on that seems maybe a bit inconsistent to me.

Reply:
I suspect this ultimately means that there needs to be a decision of some sort made here. I'm not entirely sure if this is T-lang question or a broader community one, though. Definitely does not feel like a question for two of us to make a call on.

Reply:
Some discussion on Zulip about reaching consensus here: https://zulip-archive.rust-lang.org/stream/213817-t-lang/topic/.2393678.20unused_unsafe.20changes.html.

It seems to me overall given the discussion that it is fine to merge this as proposed.